File: backend/src/services/ollama.service.ts
/**
 * Ollama Service (legacy, kept for backwards compatibility only).
 *
 * NOTE: New code should use `aiProvider` from `aiProvider.ts` instead.
 */

import axios, { AxiosInstance } from 'axios';

interface OllamaGenerateRequest {
  model: string;
  prompt: string;
  system?: string;
  stream?: boolean;
  options?: {
    temperature?: number;
    top_p?: number;
    top_k?: number;
  };
}

interface OllamaGenerateResponse {
  model: string;
  created_at: string;
  response: string;
  done: boolean;
}

interface OllamaTagsResponse {
  models: Array<{
    name: string;
    modified_at: string;
    size: number;
    digest: string;
  }>;
}

class OllamaService {
  private client: AxiosInstance;
  private baseUrl: string;
  private model: string;
  private isHealthy: boolean = false;
  private lastHealthCheck: number = 0;
  private healthCheckInterval: number = 30000; // 30 seconds

  constructor() {
    this.baseUrl = process.env.OLLAMA_URL || 'http://127.0.0.1:11434';
    this.model = process.env.OLLAMA_MODEL || 'llama3.2';
    
    this.client = axios.create({
      baseURL: this.baseUrl,
      timeout: 60000, // 60 seconds
      headers: {
        'Content-Type': 'application/json'
      }
    });

    console.log(`[OLLAMA] Service initialized with URL: ${this.baseUrl}, Model: ${this.model}`);
  }

  /**
   * Base generation method
   */
  async generate(prompt: string, systemPrompt?: string): Promise<string> {
    try {
      const request: OllamaGenerateRequest = {
        model: this.model,
        prompt,
        system: systemPrompt,
        stream: false,
        options: {
          temperature: 0.7,
          top_p: 0.9,
          top_k: 40
        }
      };

      console.log(`[OLLAMA] Generating with model: ${this.model}`);
      const response = await this.client.post<OllamaGenerateResponse>('/api/generate', request);
      
      if (response.data.done && response.data.response) {
        return response.data.response.trim();
      }
      
      throw new Error('Incomplete response from Ollama');
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.code === 'ECONNREFUSED' || error.code === 'ETIMEDOUT') {
          throw new Error('Ollama service is not available. Make sure Ollama is running.');
        }
        throw new Error(`Ollama API error: ${error.message}`);
      }
      throw error;
    }
  }

  /**
   * Improve README content
   */
  async improveReadme(readme: string, projectInfo: any): Promise<string> {
    const systemPrompt = `You are a technical documentation expert. Improve README files by adding clear descriptions, proper sections (Features, Installation, Usage, Project Structure, Contributing, License), better formatting, and professional tone. Output ONLY the improved README in markdown format, no meta-commentary.`;

    const userPrompt = `Improve this README file:

${readme}

Project Information:
- Name: ${projectInfo.name || 'Unknown'}
- Languages: ${projectInfo.languages?.join(', ') || 'Unknown'}
- Dependencies: ${projectInfo.dependencies || 0}
- Directories: ${projectInfo.directories?.join(', ') || 'None'}
- Version: ${projectInfo.version || '1.0.0'}

Provide an improved, professional README with all standard sections.`;

    return this.generate(userPrompt, systemPrompt);
  }

  /**
   * Generate project description (2-3 sentences)
   */
  async generateProjectDescription(projectInfo: any): Promise<string> {
    const systemPrompt = `You are a technical writer. Create concise project descriptions (2-3 sentences). Focus on what the project does, who it's for, and key technologies. Be specific and avoid generic phrases. Output ONLY the description, no meta-commentary.`;

    const userPrompt = `Generate a project description for:

- Name: ${projectInfo.name || 'Unknown'}
- Languages: ${projectInfo.languages?.join(', ') || 'Unknown'}
- Dependencies: ${projectInfo.dependencies || 0}
- Structure: ${projectInfo.directories?.join(', ') || 'None'}
- Type: ${projectInfo.type || 'Application'}

Create a 2-3 sentence description.`;

    return this.generate(userPrompt, systemPrompt);
  }

  /**
   * Generate setup/installation guide
   */
  async generateSetupGuide(projectInfo: any): Promise<string> {
    const systemPrompt = `You are a technical documentation expert. Create clear, step-by-step installation and setup guides. Include prerequisites, installation steps, configuration, and verification. Output ONLY the guide in markdown format.`;

    const userPrompt = `Create a setup guide for:

- Name: ${projectInfo.name || 'Project'}
- Languages: ${projectInfo.languages?.join(', ') || 'Unknown'}
- Package Manager: ${projectInfo.packageManager || 'npm'}
- Dependencies: ${projectInfo.dependencies || 0}
- Has Docker: ${projectInfo.hasDocker ? 'Yes' : 'No'}
- Has Tests: ${projectInfo.hasTests ? 'Yes' : 'No'}

Provide a complete installation and setup guide.`;

    return this.generate(userPrompt, systemPrompt);
  }

  /**
   * Generate API documentation
   */
  async generateApiDocs(endpoints: any[]): Promise<string> {
    const systemPrompt = `You are an API documentation expert. Create comprehensive API documentation with endpoint descriptions, request/response formats, examples, and error codes. Output ONLY the documentation in markdown format.`;

    const endpointsText = endpoints.map((ep, i) => 
      `${i + 1}. ${ep.method || 'GET'} ${ep.path || 'Unknown'}: ${ep.description || 'No description'}`
    ).join('\n');

    const userPrompt = `Generate API documentation for these endpoints:

${endpointsText}

Create comprehensive documentation with:
- Endpoint descriptions
- Request/response formats
- Code examples
- Error handling`;

    return this.generate(userPrompt, systemPrompt);
  }

  /**
   * Generate Mermaid diagram code
   */
  async generateMermaidDiagram(type: 'architecture' | 'dependency' | 'flow', data: any): Promise<string> {
    const systemPrompt = `You are a diagramming expert. Generate valid Mermaid diagram syntax. Output ONLY the Mermaid code, no markdown code blocks, no explanations. The output must be valid Mermaid syntax that can be directly rendered.`;

    let userPrompt = '';

    if (type === 'architecture') {
      userPrompt = `Generate a Mermaid flowchart diagram showing the architecture:

Components: ${data.components?.join(', ') || 'Unknown'}
Connections: ${JSON.stringify(data.connections || [])}
Type: ${data.type || 'flowchart'}

Create a flowchart diagram.`;
    } else if (type === 'dependency') {
      userPrompt = `Generate a Mermaid graph diagram showing dependencies:

Dependencies: ${JSON.stringify(data.dependencies || [])}
Type: ${data.type || 'graph'}

Create a dependency graph.`;
    } else {
      userPrompt = `Generate a Mermaid flowchart showing the flow:

Steps: ${data.steps?.join(' -> ') || 'Unknown'}
Type: ${data.type || 'flowchart'}

Create a flowchart.`;
    }

    const result = await this.generate(userPrompt, systemPrompt);
    // Remove markdown code blocks if present
    return result.replace(/```mermaid\n?/g, '').replace(/```\n?/g, '').trim();
  }

  /**
   * Analyze code quality and return suggestions
   */
  async analyzeCodeQuality(codeStructure: any): Promise<string[]> {
    const systemPrompt = `You are a code quality expert. Analyze code structure and provide actionable suggestions for improvement. Output ONLY a numbered list of suggestions, one per line. Be specific and practical.`;

    const userPrompt = `Analyze this code structure:

- Files: ${codeStructure.files || 0}
- Languages: ${codeStructure.languages?.join(', ') || 'Unknown'}
- Has Tests: ${codeStructure.hasTests ? 'Yes' : 'No'}
- Has Linting: ${codeStructure.hasLinting ? 'Yes' : 'No'}
- Has CI/CD: ${codeStructure.hasCI ? 'Yes' : 'No'}
- Dependencies: ${codeStructure.dependencies || 0}
- Structure: ${codeStructure.structure || 'Unknown'}

Provide 5-10 specific, actionable suggestions for improving code quality.`;

    const result = await this.generate(userPrompt, systemPrompt);
    // Parse numbered list into array
    return result
      .split('\n')
      .filter(line => line.trim().match(/^\d+[\.\)]/))
      .map(line => line.replace(/^\d+[\.\)]\s*/, '').trim())
      .filter(line => line.length > 0);
  }

  /**
   * Health check - verify Ollama is running
   */
  async healthCheck(): Promise<boolean> {
    const now = Date.now();
    
    // Cache health check for 30 seconds
    if (this.isHealthy && (now - this.lastHealthCheck) < this.healthCheckInterval) {
      return this.isHealthy;
    }

    try {
      const response = await this.client.get<OllamaTagsResponse>('/api/tags', {
        timeout: 5000 // 5 second timeout for health check
      });
      
      this.isHealthy = response.status === 200;
      this.lastHealthCheck = now;
      return this.isHealthy;
    } catch (error) {
      this.isHealthy = false;
      this.lastHealthCheck = now;
      console.warn('[OLLAMA] Health check failed:', error instanceof Error ? error.message : 'Unknown error');
      return false;
    }
  }

  /**
   * List available models
   */
  async listModels(): Promise<string[]> {
    try {
      const response = await this.client.get<OllamaTagsResponse>('/api/tags');
      return response.data.models.map(model => model.name);
    } catch (error) {
      console.error('[OLLAMA] Failed to list models:', error);
      return [];
    }
  }
}

// Export singleton instance
export const ollamaService = new OllamaService();

File: backend/services/ollama_service.py
"""Ollama Service for AI-powered features."""
import httpx
import json
from typing import Optional, List, Dict, Any
import os
import traceback

class OllamaService:
    def __init__(self):
        self.base_url = os.getenv('OLLAMA_URL', 'http://127.0.0.1:11434')
        self.model = os.getenv('OLLAMA_MODEL', 'llama3.2')
        self.timeout = 60.0  # 60 seconds
        print("=" * 80)
        print(f"[OllamaService] [OK] Initialized with URL: {self.base_url}, Model: {self.model}")
        print("=" * 80)
    
    async def generate(self, prompt: str, system_prompt: Optional[str] = None) -> str:
        """Generate text using Ollama"""
        try:
            print(f"[Ollama] Generating with prompt length: {len(prompt)} chars")
            if system_prompt:
                print(f"[Ollama] System prompt length: {len(system_prompt)} chars")
            
            async with httpx.AsyncClient(timeout=self.timeout) as client:
                payload = {
                    'model': self.model,
                    'prompt': prompt,
                    'stream': False,
                    'options': {
                        'temperature': 0.7,
                        'top_p': 0.9,
                        'top_k': 40
                    }
                }
                
                if system_prompt:
                    payload['system'] = system_prompt
                
                print(f"[Ollama] Sending request to: {self.base_url}/api/generate")
                print(f"[Ollama] Model: {self.model}, Stream: False")
                
                response = await client.post(
                    f'{self.base_url}/api/generate',
                    json=payload
                )
                
                print(f"[Ollama] Response status: {response.status_code}")
                
                if response.status_code != 200:
                    error_text = response.text[:500]  # Limit error text length
                    print(f"[Ollama] [ERROR] Error response: {error_text}")
                    raise Exception(f'Ollama returned status {response.status_code}: {error_text}')
                
                data = response.json()
                result = data.get('response', '').strip()
                
                if not result:
                    print(f"[Ollama] [ERROR] Empty response from Ollama. Full data: {data}")
                    raise Exception('Ollama returned empty response')
                
                print(f"[Ollama] [OK] Generation complete. Response length: {len(result)} chars")
                return result
                
        except httpx.TimeoutException as e:
            print(f"[Ollama] [ERROR] Timeout error: {e}")
            traceback.print_exc()
            raise Exception('Ollama request timed out. The model may be loading or too slow.')
        except httpx.ConnectError as e:
            print(f"[Ollama] [ERROR] Connection error: {e}")
            traceback.print_exc()
            raise Exception(f'Cannot connect to Ollama at {self.base_url}. Make sure Ollama is running.')
        except Exception as e:
            print(f"[Ollama] [ERROR] Generation error: {type(e).__name__}: {str(e)}")
            traceback.print_exc()
            raise Exception(f'Failed to generate content: {str(e)}')
    
    async def improve_readme(self, readme: str, project_info: Dict[str, Any]) -> str:
        """Improve README content with AI"""
        print("=" * 80)
        print("[Ollama] improve_readme called")
        print(f"[Ollama] README length: {len(readme)} chars")
        print(f"[Ollama] Project info: {json.dumps(project_info, indent=2)}")
        
        system_prompt = """You are a senior technical documentation expert. Improve README files by:
- Adding clear, concise descriptions
- Organizing content logically
- Adding proper sections (Features, Installation, Usage, etc.)
- Improving formatting and readability
- Using professional, friendly tone

Output ONLY the improved README in markdown format. Do not add explanations or meta-commentary."""

        languages = ', '.join(project_info.get('languages', ['Unknown']))
        prompt = f"""Improve this README for a {languages} project:

Project Name: {project_info.get('name', 'Unknown')}
Current README:
{readme or 'No README found'}

Additional Context:
- Dependencies: {project_info.get('dependencies', 'Unknown')}
- Main Directories: {', '.join(project_info.get('directories', []))}
- Version: {project_info.get('version', 'Unknown')}

Generate an improved README with these sections:
1. Project title and brief description
2. Features (based on directory structure)
3. Prerequisites
4. Installation
5. Usage
6. Project Structure
7. Contributing (if appropriate)
8. License

Make it professional, clear, and actionable."""

        try:
            result = await self.generate(prompt, system_prompt)
            print(f"[Ollama] [OK] improve_readme completed. Result length: {len(result)} chars")
            print("=" * 80)
            return result
        except Exception as e:
            print(f"[Ollama] [ERROR] improve_readme failed: {e}")
            traceback.print_exc()
            print("=" * 80)
            raise
    
    async def generate_project_description(self, project_info: Dict[str, Any]) -> str:
        """Generate a concise project description"""
        system_prompt = """You are a technical writer. Create concise project descriptions (2-3 sentences).
Focus on what the project does, who it's for, and key technologies. Be specific and avoid generic phrases.
Output ONLY the description, nothing else."""

        prompt = f"""Create a project description for:

Name: {project_info.get('name', 'Unknown')}
Languages: {', '.join(project_info.get('languages', []))}
Dependencies: {project_info.get('dependencies', 'Unknown')}
Directories: {', '.join(project_info.get('directories', []))}
Version: {project_info.get('version', 'Unknown')}"""

        return await self.generate(prompt, system_prompt)
    
    async def generate_setup_guide(self, project_info: Dict[str, Any]) -> str:
        """Generate setup/installation guide"""
        system_prompt = """You are a technical documentation expert. Create clear, step-by-step installation and setup guides.
Include prerequisites, installation steps, configuration, and verification. Output ONLY the guide in markdown format."""

        prompt = f"""Create a setup guide for:

- Name: {project_info.get('name', 'Project')}
- Languages: {', '.join(project_info.get('languages', []))}
- Package Manager: {project_info.get('packageManager', 'npm')}
- Dependencies: {project_info.get('dependencies', 0)}
- Has Docker: {project_info.get('hasDocker', False)}
- Has Tests: {project_info.get('hasTests', False)}

Provide a complete installation and setup guide."""

        return await self.generate(prompt, system_prompt)
    
    async def generate_api_docs(self, endpoints: List[Dict[str, Any]]) -> str:
        """Generate API documentation"""
        system_prompt = """You are an API documentation expert. Create comprehensive API documentation with endpoint descriptions, request/response formats, examples, and error codes. Output ONLY the documentation in markdown format."""

        endpoints_text = '\n'.join([
            f"{i+1}. {ep.get('method', 'GET')} {ep.get('path', 'Unknown')}: {ep.get('description', 'No description')}"
            for i, ep in enumerate(endpoints)
        ])

        prompt = f"""Generate API documentation for these endpoints:

{endpoints_text}

Create comprehensive documentation with:
- Endpoint descriptions
- Request/response formats
- Code examples
- Error handling"""

        return await self.generate(prompt, system_prompt)
    
    async def generate_mermaid_diagram(self, diagram_type: str, data: Dict[str, Any]) -> str:
        """Generate Mermaid diagram code"""
        print("=" * 80)
        print(f"[Ollama] generate_mermaid_diagram called with type: {diagram_type}")
        print(f"[Ollama] Data: {json.dumps(data, indent=2)}")
        
        system_prompt = """You are a diagramming expert. Generate valid Mermaid diagram syntax.
Output ONLY the Mermaid code, no markdown code blocks, no explanations. The output must be valid Mermaid syntax that can be directly rendered."""

        prompts = {
            'architecture': f"""Generate a Mermaid flowchart diagram (flowchart TB) showing the architecture:

Components: {', '.join(data.get('components', []))}
Connections: {json.dumps(data.get('connections', []), indent=2)}

Create a flowchart diagram showing the system architecture.""",
            
            'dependency': f"""Generate a Mermaid graph diagram showing dependencies:

Dependencies: {json.dumps(data.get('dependencies', []), indent=2)}

Create a dependency graph.""",
            
            'flow': f"""Generate a Mermaid flowchart showing the flow:

Steps: {' -> '.join(data.get('steps', []))}

Create a flowchart."""
        }

        prompt = prompts.get(diagram_type, prompts['architecture'])
        
        try:
            result = await self.generate(prompt, system_prompt)
            # Remove markdown code blocks if present
            cleaned = result.replace('```mermaid\n', '').replace('```\n', '').replace('```', '').strip()
            print(f"[Ollama] [OK] generate_mermaid_diagram completed. Result length: {len(cleaned)} chars")
            print("=" * 80)
            return cleaned
        except Exception as e:
            print(f"[Ollama] [ERROR] generate_mermaid_diagram failed: {e}")
            traceback.print_exc()
            print("=" * 80)
            raise
    
    async def analyze_code_quality(self, code_structure: Dict[str, Any]) -> List[str]:
        """Analyze code quality and return suggestions"""
        system_prompt = """You are a code quality expert. Analyze code structure and provide actionable suggestions for improvement. Output ONLY a numbered list of suggestions, one per line. Be specific and practical."""

        prompt = f"""Analyze this code structure:

- Files: {code_structure.get('files', 0)}
- Languages: {', '.join(code_structure.get('languages', []))}
- Has Tests: {code_structure.get('hasTests', False)}
- Has Linting: {code_structure.get('hasLinting', False)}
- Has CI/CD: {code_structure.get('hasCI', False)}
- Dependencies: {code_structure.get('dependencies', 0)}
- Structure: {code_structure.get('structure', 'Unknown')}

Provide 5-10 specific, actionable suggestions for improving code quality."""

        result = await self.generate(prompt, system_prompt)
        # Parse numbered list into array
        import re
        suggestions = []
        for line in result.split('\n'):
            line = line.strip()
            # Match numbered items (1., 2), etc.)
            match = re.match(r'^\d+[\.\)]\s*(.+)', line)
            if match:
                suggestions.append(match.group(1).strip())
        return [s for s in suggestions if len(s) > 0]
    
    async def health_check(self) -> bool:
        """Check if Ollama is running"""
        try:
            print(f"[Ollama] Health check: Checking {self.base_url}/api/tags")
            async with httpx.AsyncClient(timeout=5.0) as client:
                response = await client.get(f'{self.base_url}/api/tags')
                is_healthy = response.status_code == 200
                status_msg = '[OK] Healthy' if is_healthy else f'[ERROR] Unhealthy (status: {response.status_code})'
                print(f"[Ollama] Health check: {status_msg}")
                return is_healthy
        except Exception as e:
            print(f"[Ollama] [ERROR] Health check failed: {type(e).__name__}: {str(e)}")
            traceback.print_exc()
            return False
    
    async def list_models(self) -> List[str]:
        """List available Ollama models"""
        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                response = await client.get(f'{self.base_url}/api/tags')
                if response.status_code == 200:
                    data = response.json()
                    return [model['name'] for model in data.get('models', [])]
                return []
        except:
            return []

# Singleton instance
ollama_service = OllamaService()
print("[OllamaService] [OK] Singleton instance created")